storage/sql/template.go

```diff
@@ -0,0 +1,401 @@
+		if offsetValue.GetType() == octosql.TypeNull { // the variable doesn't exist, merge it with variables
+			offsetVariable := octosql.NewVariables(map[octosql.VariableName]octosql.Value{offsetPlaceholderName: newOffsetValue})
+			rs.variables, err = rs.variables.MergeWith(offsetVariable)
+			if err != nil {
+				log.Fatalf("sql worker: couldn't merge variables with offset variable: %s", err)
+				return
+			}
+		} else { // just update value in variables // TODO - maybe think about some Set() method on variables? MergeWith returns error when exists
+			rs.variables[offsetPlaceholderName] = newOffsetValue
+		}
+
+		values := make([]interface{}, 0)
+
+		for _, expression := range rs.placeholders {
+			// Since we have an execution expression, then we can evaluate it given the variables
+			value, err := expression.ExpressionValue(ctx, rs.variables)
+			if err != nil {
+				log.Fatalf("sql worker: couldn't get actual value from variables: %s", err)
+				return
+			}
+
+			values = append(values, value.ToRawValue())
+		}
+
+		rows, err := rs.stmt.QueryContext(ctx, values...)
+		if err != nil {
+			log.Fatalf("sql worker: couldn't query statement: %s", err)
+			return
+		}
+		rs.rows = rows
+
+		columns, err := rows.Columns()
+		if err != nil {
+			log.Fatalf("sql worker: couldn't get columns from rows: %s", err)
+			return
+		}
+		rs.columns = columns
+
+		for { // inner for is calling RunWorkerInternal
+			select {
+			case <-ctx.Done():
+				rs.workerCloseErrChan <- ctx.Err()
+			default:
+			}
+
+			tx := rs.stateStorage.BeginTransaction().WithPrefix(rs.streamID.AsPrefix())
+
+			err := rs.RunWorkerInternal(ctx, tx)
+			if errors.Cause(err) == execution.ErrNewTransactionRequired {
+				tx.Abort()
+				continue
+			} else if waitableError := execution.GetErrWaitForChanges(err); waitableError != nil {
+				tx.Abort()
+				err = waitableError.ListenForChanges(ctx)
+				if err != nil {
+					log.Println("sql worker: couldn't listen for changes: ", err)
+				}
+				err = waitableError.Close()
+				if err != nil {
+					log.Println("sql worker: couldn't close storage changes subscription: ", err)
+				}
+				continue
+			} else if err == execution.ErrEndOfStream {
+				err = tx.Commit()
+				if err != nil {
+					log.Println("sql worker: couldn't commit transaction: ", err)
+					continue
+				}
+				return
+			} else if err != nil {
+				tx.Abort()
+				log.Printf("sql worker: error running sql read batch worker: %s, reinitializing from storage", err)
+				break
+			}
+
+			err = tx.Commit()
+			if err != nil {
+				log.Println("sql worker: couldn't commit transaction: ", err)
+				continue
+			}
+		}
+	}
+}
+
+var outputQueuePrefix = []byte("$output_queue$")
+
+func (rs *RecordStream) RunWorkerInternal(ctx context.Context, tx storage.StateTransaction) error {
+	outputQueue := execution.NewOutputQueue(tx.WithPrefix(outputQueuePrefix))
+
+	if rs.isDone {
+		err := outputQueue.Push(ctx, &kafka.QueueElement{
```

Comment:
You shouldn't have a dependency on the kafka package here. Create your own queue element.